File: py/aiomavsdk/examples/telemetry.py
import asyncio
from aiomavsdk import Mavsdk, Configuration, ComponentType
from aiomavsdk.plugins.action import ActionAsync
from aiomavsdk.plugins.telemetry import TelemetryAsync, LandedState


async def print_positions(telemetry: TelemetryAsync):
    async for position in telemetry.subscribe_position():
        print(f"Position update: Lat={position.latitude_deg:.6f}, Lon={position.longitude_deg:.6f}, Alt={position.relative_altitude_m:.1f}m")


async def wait_for_landing(telemetry: TelemetryAsync):
    async for state in telemetry.subscribe_landed_state():
        print(f"Landed state: {state}")
        if state == LandedState.ON_GROUND:
            return


async def main():
    configuration = Configuration.create_with_component_type(ComponentType.GROUND_STATION)
    mavsdk = Mavsdk(configuration)
    await mavsdk.add_any_connection("udpin://0.0.0.0:14540")

    print("Waiting for a drone...")

    drone = None
    async for _ in mavsdk.on_new_system():
        systems = await mavsdk.get_systems()
        for system in systems:
            if await system.has_autopilot() and await system.is_connected():
                print("✓ Found valid autopilot system!")
                drone = system
                break
        if drone is not None:
            break

    if not drone:
        print("No valid system found!")
        return

    action = ActionAsync(drone)
    telemetry = TelemetryAsync(drone)

    position_task = asyncio.create_task(print_positions(telemetry))

    print("Arming...")
    await action.arm()

    print("Taking off...")
    await action.takeoff()

    print("Flying for 10 seconds...")
    await asyncio.sleep(10)

    print("Current position (sync poll):")
    current_pos = await telemetry.position()
    print(f"  Lat: {current_pos.latitude_deg:.6f}")
    print(f"  Lon: {current_pos.longitude_deg:.6f}")
    print(f"  Alt: {current_pos.relative_altitude_m:.1f}m")

    print("Landing...")
    await action.land()

    print("Waiting for landing completion...")
    await wait_for_landing(telemetry)

    print("Mission complete!")
    position_task.cancel()


if __name__ == "__main__":
    asyncio.run(main())

File: kt/examples/cli-jvm/src/main/kotlin/io/mavsdk/kotlin/examples/Telemetry.kt
package io.mavsdk.kotlin.examples

import io.mavsdk.kotlin.ComponentType
import io.mavsdk.kotlin.Mavsdk
import io.mavsdk.kotlin.plugins.action.Action
import io.mavsdk.kotlin.plugins.telemetry.Telemetry
import kotlinx.coroutines.cancelAndJoin
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.collect
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.launch
import kotlinx.coroutines.runBlocking

fun telemetry() = runBlocking {
    println("=== MAVSDK-Kotlin Telemetry Example ===\n")

    Mavsdk(ComponentType.GROUND_STATION).use { mavsdk ->
        mavsdk.addAnyConnection("udpin://0.0.0.0:14540")
            .onFailure {
                println("Connection failed: $it")
                return@runBlocking
            }

        println("Waiting for a drone...")
        val system = mavsdk.firstAutopilot(timeoutSeconds = 10.0) ?: run {
            println("No valid autopilot found")
            return@runBlocking
        }

        val action = Action.create(system)
        val telemetry = Telemetry.create(system)
        val positionJob = launch {
            telemetry.subscribePosition().collect { position ->
                println(
                    "Position update: Lat=%.6f, Lon=%.6f, Alt=%.1fm".format(
                        position.latitudeDeg,
                        position.longitudeDeg,
                        position.relativeAltitudeM,
                    )
                )
            }
        }

        println("Arming...")
        check(action.arm() == Action.Result.SUCCESS) { "Arming failed" }

        println("Taking off...")
        check(action.takeoff() == Action.Result.SUCCESS) { "Takeoff failed" }

        println("Flying for 10 seconds...")
        delay(10_000)

        val currentPosition = telemetry.position()
        println("Current position:")
        println("  Lat: %.6f".format(currentPosition.latitudeDeg))
        println("  Lon: %.6f".format(currentPosition.longitudeDeg))
        println("  Alt: %.1fm".format(currentPosition.relativeAltitudeM))

        println("Landing...")
        check(action.land() == Action.Result.SUCCESS) { "Landing failed" }

        println("Waiting for landing completion...")
        telemetry.subscribeLandedState().first { state ->
            println("Landed state: $state")
            state == Telemetry.LandedState.ON_GROUND
        }

        positionJob.cancelAndJoin()
        println("Flight complete")
    }
}
